motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,435 @@
+// Separating this allows other things to call rrtBackgroundRunner in parallel allowing the thread-agnostic Plan to be accessible.
+func (mp *tpspaceRRTMotionPlanner) rrtBackgroundRunner(
+	ctx context.Context,
+	_ []referenceframe.Input,
+	rrt *rrtParallelPlannerShared,
+) {
+	defer close(rrt.solutionChan)
+
+	tpFrame, ok := mp.frame.(tpspace.PTGProvider)
+	if !ok {
+		rrt.solutionChan <- &rrtPlanReturn{planerr: fmt.Errorf("frame %v must be a PTGProvider", mp.frame)}
+		return
+	}
+
+	// get start and goal poses
+	var startPose spatialmath.Pose
+	var goalPose spatialmath.Pose
+	for k := range rrt.maps.startMap {
+		if cNode, ok := k.(*configurationNode); ok {
+			startPose = cNode.endConfig
+		} else {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: fmt.Errorf("node %v must be a configurationNode", k)}
+			return
+		}
+	}
+	for k := range rrt.maps.goalMap {
+		if cNode, ok := k.(*configurationNode); ok {
+			goalPose = cNode.endConfig
+		} else {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: fmt.Errorf("node %v must be a configurationNode", k)}
+			return
+		}
+	}
+
+	dist := math.Sqrt(mp.planOpts.DistanceFunc(&Segment{StartPosition: startPose, EndPosition: goalPose}))
+	midPt := goalPose.Point().Sub(startPose.Point())
+
+	success := false
+	var successNode *configurationNode
+	iter := 0
+	// While not at goal:
+	// TODO: context, timeout, etc
+	for !success {
+		select {
+		case <-ctx.Done():
+			mp.logger.Debugf("TP Space RRT timed out after %d iterations", iter)
+			rrt.solutionChan <- &rrtPlanReturn{planerr: fmt.Errorf("TP Space RRT timeout %w", ctx.Err()), maps: rrt.maps}
+			return
+		default:
+		}
+		// Get random cartesian configuration
+		randPos := goalPose
+		tryGoal := true
+		// Check if we can reach the goal every N iters
+		if iter%mp.goalCheck != 0 {
+			rDist := dist * (mp.autoBB + float64(iter)/10.)
+			tryGoal = false
+			randPosX := float64(mp.randseed.Intn(int(rDist)))
+			randPosY := float64(mp.randseed.Intn(int(rDist)))
+			randPosTheta := math.Pi * (mp.randseed.Float64() - 0.5)
+			randPos = spatialmath.NewPose(
+				r3.Vector{midPt.X + (randPosX - rDist/2.), midPt.Y + (randPosY - rDist/2.), 0},
+				&spatialmath.OrientationVector{OZ: 1, Theta: randPosTheta},
+			)
+		}
+		iter++
+		randPosNode := &configurationNode{endConfig: randPos}
+
+		candidateNodes := map[float64][2]*configurationNode{}
+
+		// Find the best traj point for each traj family, and store for later comparison
+		// TODO: run in parallel
+		for ptgNum, curPtg := range tpFrame.PTGs() {
+			cand := mp.getExtensionCandidate(ctx, tryGoal, randPosNode, ptgNum, curPtg, rrt)
+			if cand != nil {
+				if cand.finalGoal {
+					// If we've reached the goal, break out
+					// TODO: do this better
+					success = true
+					rrt.maps.startMap[cand.newNode] = cand.treeNode
+					successNode = cand.newNode
+					candidateNodes = nil
+					break
+				}
+				candidateNodes[cand.dist] = [2]*configurationNode{cand.treeNode, cand.newNode}
+			}
+		}
+		if !success {
```

Comment:
Actually should not need this if statement.  You break from the loop if its a success